Post Editor: reduxify setting the featured image

jsnajdr opened the pull request:
This PR reduxifies the UI to set post featured image. It can be edited at two places: the area above the text editor, and in a dedicated section in the editor drawer:

<img width="982" alt="screen shot 2018-04-09 at 15 16 27" src="https://user-images.githubusercontent.com/664258/38499890-373ad0b6-3c09-11e8-994d-48fdc580758a.png">

**How to test:**
With an existing post with a featured image, check that the featured image is shown after opening in editor.

While editing a post, try to add, change or remove a featured image. Verify that the changes get saved.

Start editing a new post, and without doing any other changes, set the feature image and then immediately remove it. Click on "Close" in the top left corner, i.e., discard the draft without saving. Verify that the editor is closed without asking if you want to save changes (there are no changes). In other words, verify that the fix #17701 has not regressed. Instead of checking inside the `set` function in Post Edit Store, we now set a correct value of `DEFAULT_NEW_POST_VALUES.featured_image` to achieve the same behavior.

Start editing a new post, and in "More Options", select a post to copy. Verify that the copied data include the featured image.

jsnajdr committed:
Post Editor: reduxify setting the featured image

samouri commented:
> In other words, verify that the fix #17701 has not regressed. Instead of checking inside the set function in Post Edit Store, we now set a correct value of DEFAULT_NEW_POST_VALUES.featured_image to achieve the same behavior.

awesome!

samouri commented on client/post-editor/editor-featured-image/dropzone.jsx:
```diff
@@ -67,21 +66,7 @@ class FeaturedImageDropZone extends Component {
 			}
 
-			/**
-			 * TODO: Redux way. What's left: figure out how to properly use `editPost`
```
todo is to-done!

samouri commented:
its suspicious that an e2e test failed specifically about publishing posts. rerunning now

samouri commented:
i followed the testing instructions and everything checked out wonderfully.
the only bit i found that seemed wrong was this flow:

1. add a featured image
2. click "update" or "save"
3. when you try to close the post, you will get the unsaved changes dialog even though you just saved it!

jsnajdr commented:
> when you try to close the post, you will get the unsaved changes dialog even though you just saved it!

Great catch! That's really a bug. I'll need to find a way how to fix it.

jsnajdr committed:
Migrate post utils tests to Jest asserts

jsnajdr committed:
Post Editor: Correctly detect saved and noop edits of featured_image property

The `featured_image` property has quite a different meaning when sending a save
request and when receiving a post. Compare them correctly in the reducer when
receiving edits and in the `isEditedPostDirty` selector.

jsnajdr commented:
> when you try to close the post, you will get the unsaved changes dialog even though you just saved it!

@samouri I pushed a fix for this issue. The cause is that the `featured_image` property has quite a different meaning when saving a post and when receiving post data.

When saving the post, we send a POST body where the property value is the ID of the image in the media library:
```js
{
  featured_image: 123
}
```

To remove the featured image, we send an empty string:
```js
{
  featured_image: ''
}
```

When receiving a post, the `featured_image` value is an URL and the image ID is a property of the `post_thumbnail` object:
```js
{
  featured_image: 'https://example.files.wordpress.com/...',
  post_thumbnail: {
    ID: 123,
    URL: ...,
    width: ...,
    height, ...
  }
}
```

And this is how a post without a featured image looks like:
```js
{
  featured_image: '',
  post_thumbnail: null
}
```

There is a `getFeaturedImageId( post )` helper function to read the right value from a `post` object. I updated the `edits` reducer and the `isEditedPostDirty` selector to use it to compare the values correctly. The bug you reported got fixed.

In a drive-by commit, I updated the `lib/posts` tests to use Jest asserts instead of Node `assert` module.

samouri reviewed (approved):
this was some awesome work ⭐️

gwwar commented on client/post-editor/post-editor.jsx:
```diff
@@ -360,12 +360,7 @@ export const PostEditor = createReactClass( {
-									maxWidth={ 1462 }
-									hasDropZone={ true }
-								/>
+								<FeaturedImage maxWidth={ 1462 } hasDropZone />
```
Total aside that this magic number seems quite odd. 🤷‍♀️  Doing a quick search in history I didn't find a reason why it's this value.

jsnajdr replied:
It's been this way since the initial OSS commit. I also don't know why this precise value :smile:

gwwar reviewed (approved):
Awesome work! This tests well for me.

Only found one minor thing. While testing through steps in https://github.com/Automattic/wp-calypso/issues/17701 I did notice that we get an unsaved changes notice for published posts. 
1. edit a published post
2. set a featured image
3. remove the featured image
4. Close the editor

This is pretty minor though and we can address it in a follow up if other folks can reproduce.
